internal/provider_framework/attribute_validators.go

```diff
@@ -0,0 +1,368 @@
+}
+
+// conflictsWith is a helper to instantiate conflictsWithAttributeValidator.
+func conflictsWith(attributePaths ...*tftypes.AttributePath) tfsdk.AttributeValidator {
+	return &conflictsWithAttributeValidator{attributePaths}
+}
+
+var _ tfsdk.AttributeValidator = (*conflictsWithAttributeValidator)(nil)
+
+func (av conflictsWithAttributeValidator) Description(ctx context.Context) string {
+	return av.MarkdownDescription(ctx)
+}
+
+func (av conflictsWithAttributeValidator) MarkdownDescription(_ context.Context) string {
+	return fmt.Sprintf("Ensure that if an attribute is set, these are not set: %q", av.attrPaths)
+}
+
+func (av conflictsWithAttributeValidator) Validate(ctx context.Context, req tfsdk.ValidateAttributeRequest, res *tfsdk.ValidateAttributeResponse) {
+	tflog.Debug(ctx, "Validating attribute is not set together with other conflicting attributes", map[string]interface{}{
+		"attribute":             attrPathToString(req.AttributePath),
+		"conflictingAttributes": av.attrPaths,
+	})
+
+	var v attr.Value
+	res.Diagnostics.Append(tfsdk.ValueAs(ctx, req.AttributeConfig, &v)...)
+	if res.Diagnostics.HasError() {
+		return
+	}
+
+	for _, path := range av.attrPaths {
+		var o attr.Value
+		res.Diagnostics.Append(req.Config.GetAttribute(ctx, path, &o)...)
+		if res.Diagnostics.HasError() {
+			return
+		}
+
+		if !v.IsNull() && !o.IsNull() {
+			res.Diagnostics.AddAttributeError(
+				req.AttributePath,
+				fmt.Sprintf("Attribute %q conflicting", attrPathToString(path)),
+				fmt.Sprintf("%q cannot be specified when %q is specified", attrPathToString(path), attrPathToString(req.AttributePath)),
+			)
+			return
+		}
+	}
+}
+
+// exactlyOneOfAttributeValidator checks that of a set of *tftypes.AttributePath,
+// including the attribute it's applied to, one and only one attribute out of all specified is configured.
+// It will also cause a validation error if none are specified.
+//
+// The provided tftypes.AttributePath must be "absolute",
+// and starting with top level attribute names.
+type exactlyOneOfAttributeValidator struct {
+	attrPaths []*tftypes.AttributePath
+}
+
+// exactlyOneOf is a helper to instantiate exactlyOneOfAttributeValidator.
+func exactlyOneOf(attributePaths ...*tftypes.AttributePath) tfsdk.AttributeValidator {
+	return &exactlyOneOfAttributeValidator{attributePaths}
+}
+
+var _ tfsdk.AttributeValidator = (*exactlyOneOfAttributeValidator)(nil)
+
+func (av exactlyOneOfAttributeValidator) Description(ctx context.Context) string {
+	return av.MarkdownDescription(ctx)
+}
+
+func (av exactlyOneOfAttributeValidator) MarkdownDescription(_ context.Context) string {
+	return fmt.Sprintf("Ensure that one and only one attribute from this collection is set: %q", av.attrPaths)
+}
+
+func (av exactlyOneOfAttributeValidator) Validate(ctx context.Context, req tfsdk.ValidateAttributeRequest, res *tfsdk.ValidateAttributeResponse) {
+	// Assemble a slice of paths, ensuring we don't repeat the attribute this validator is applied to
```

Comment:
Is a check for `if req.AttributeConfig.IsNull() || req.AttributeConfig.IsUnknown()` required here?
If not, should there be one for https://github.com/hashicorp/terraform-provider-tls/pull/215/files#diff-c32022ba2194630d49dde2e8e91220e983a86076686d51bbc96f3edc6aba6f29R117 ?

Reply:
https://github.com/hashicorp/terraform-provider-tls/pull/215#discussion_r882099448